src/Disks/IO/ReadBufferFromRemoteFSGather.h

```diff
@@ -67,20 +67,24 @@ friend class ReadIndirectBufferFromRemoteFS;
 
 
 #if USE_AWS_S3
+class DiskCache;
+
 /// Reads data from S3 using stored paths in metadata.
 class ReadBufferFromS3Gather final : public ReadBufferFromRemoteFSGather
```

Comment:
btw why not just put all this logic with [read from cache / wait until downloaded / read without cache] inside already existing `ReadBufferFromRemoteFSGather`? Because making all methods of `ReadBuffer` as virtual seems like a bad idea, I am sure nobody will allow doing that.

btw in this PR https://github.com/ClickHouse/ClickHouse/pull/31104 a person also made local cache, but not on disk interface level, but on the level of table engine (for now it is only Hive table engine, but I will not merge his PR until this local caching part is generalized enough to be easily applied to other integration table engines like s3 and hdfs). 
In his PR  same as in your PR, he adds reading with exact ranges with both range boundaries:https://github.com/ClickHouse/ClickHouse/blob/b28f3a2ad413efa748500dd3b2e4c4489556741f/src/IO/RemoteReadBufferCache.cpp#L100
So if he did not need to make all methods of `ReadBuffer` as virtual so you do not need either.
His PR is far not perfect now, but has some good ideas.
For example he has a good (at least for me) idea of having a singleton class and accessing it like `RemoteReadBufferCache::instance()`. It simplifies the code.

But actually may be it could be better to move that logic not inside `ReadBufferFromRemoteFSGather` but to make an additional buffer between `ReadBufferFromRemoteFSGather` and `ReadBufferFromS3`. (Though here need to be careful not forget that it requires some small adjustments not to break asynchronous reads feature -- just to set correct working_buffer under condition `use_external_buffer` and to add methods `setReadUntilEnd` and `setReadUntilPosition`).

What do you think?

Reply:
In this variant new virtual methods still be virtual. I need to change them to return from different internal buffers (S3 or local) or chane behavior for methods like set().